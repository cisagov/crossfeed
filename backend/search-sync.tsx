import { Domain, connectToDatabase } from '../models';
import { CommandOptions } from './ecs-client';
import * as wappalyzer from 'simple-wappalyzer';
import Apify, { CheerioHandlePage } from 'apify';
import { Client } from '@elastic/elasticsearch';
import * as crypto from 'crypto';

const INDEX_NAME = 'webpages';

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName, scanTaskId } = commandOptions;

  console.log('Running crawl on organization', organizationName);
  await connectToDatabase();
  const domains = await Domain.find({
    where: { organization: { id: organizationId } },
    select: ['id', 'name', 'ip']
  });

  const client = new Client({ node: process.env.ELASTICSEARCH_ENDPOINT! });

  process.env.APIFY_LOCAL_STORAGE_DIR =
    __dirname + '/apify-cache/' + Math.random();

  const requestQueue = await Apify.openRequestQueue();
  for (const domain of domains) {
    const userData = {
      domain
    };
    for (const url of [
      `http://${domain.name}`,
      `http://${domain.name}/robots.txt`,
      `http://${domain.name}/sitemap.xml`
    ]) {
      await requestQueue.addRequest({ url: `${url}`, userData });
    }
  }
  const dataset = await Apify.openDataset('pages');
  const handlePageFunction: CheerioHandlePage = async ({
    request,
    response,
    body,
    $
  }) => {
    console.log(request.url);
    const webTechnologies = await wappalyzer({
      url: response.url,
      data: body,
      statusCode: response.statusCode,
      headers: response.headers
    });
    const item = {
      domainName: request.userData.domain.name,
      domainId: request.userData.domain.id,
      organizationId: organizationId,
      organizationName: organizationName,
      scanTaskId: scanTaskId,
      timestamp: new Date(),
      url: request.url,
      body: body,
      text: Apify.utils.htmlToText($ as CheerioStatic),
      webTechnologies,
      statusCode: response.statusCode
    };
    await dataset.pushData(item);
    // Add all links from page to ReqxwuestQueue
    await Apify.utils.enqueueLinks({
      $: $ as CheerioStatic,
      requestQueue,
      baseUrl: request.loadedUrl,
      transformRequestFunction: (e) => ({
        ...e,
        userData: request.userData
      }),
      pseudoUrls: [`http[s?]://${request.userData.domain.name}/[.+]`]
    });
  };
  // Create a CheerioCrawler
  const crawler = new Apify.CheerioCrawler({
    requestQueue,
    handlePageFunction,
    maxRequestsPerCrawl: 100,
    maxConcurrency: 10
  });
  // Run the crawler
  await crawler.run();

  const data = await dataset.getData();
  console.error(data.items.map((e) => (e as any).url));

  await client.helpers.bulk({
    datasource: data.items,
    onDocument(doc: any) {
      return [
        {
          update: {
            _index: INDEX_NAME,
            _id: crypto
              .createHash('sha1')
              .update(doc.url)
              .update(doc.domainId)
              .update(doc.organizationId)
              .digest('base64')
          }
        },
        { doc_as_upsert: true }
      ];
    }
  });
  /**
backend_1   | {
backend_1   |   took: 34,
backend_1   |   errors: false,
backend_1   |   items: [
backend_1   |     { index: [Object] },
backend_1   |     { index: [Object] },
backend_1   |     { index: [Object] },
backend_1   |     { index: [Object] }
backend_1   |   ]
backend_1   | }
     */
};
