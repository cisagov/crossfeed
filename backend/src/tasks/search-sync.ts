import { Domain, connectToDatabase, Vulnerability, Webpage } from '../models';
import { CommandOptions } from './ecs-client';
import { In } from 'typeorm';
import ESClient from './es-client';
import { chunk } from 'lodash';
import pRetry from 'p-retry';

/**
 * Chunk sizes. These values are small during testing to facilitate testing.
 */
export const DOMAIN_CHUNK_SIZE = typeof jest === 'undefined' ? 100 : 10;

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, domainId } = commandOptions;

  console.log('Running searchSync');
  await connectToDatabase();

  const client = new ESClient();

  const qs = Domain.createQueryBuilder('domain')
    .leftJoinAndSelect('domain.organization', 'organization')
    .leftJoinAndSelect('domain.vulnerabilities', 'vulnerabilities')
    .leftJoinAndSelect('domain.services', 'services')
    .having('domain.syncedAt is null')
    .orHaving('domain.updatedAt > domain.syncedAt')
    .orHaving('organization.updatedAt > domain.syncedAt')
    .orHaving(
      'COUNT(CASE WHEN vulnerabilities.updatedAt > domain.syncedAt THEN 1 END) >= 1'
    )
    .orHaving(
      'COUNT(CASE WHEN services.updatedAt > domain.syncedAt THEN 1 END) >= 1'
    )
    .groupBy('domain.id, organization.id, vulnerabilities.id, services.id')
    .select(['domain.id']);

  if (organizationId) {
    // This parameter is used for testing only
    qs.where('organization.id=:org', { org: organizationId });
  }

  const domainIds = (await qs.getMany()).map((e) => e.id);
  console.log(`Got ${domainIds.length} domains.`);
  if (domainIds.length) {
    const domainIdChunks = chunk(domainIds, DOMAIN_CHUNK_SIZE);
    for (const domainIdChunk of domainIdChunks) {
      const domains = await Domain.find({
        where: { id: In(domainIdChunk) },
        relations: ['services', 'organization', 'vulnerabilities']
      });
      console.log(`Syncing ${domains.length} domains...`);
      await pRetry(() => client.updateDomains(domains), {
        retries: 3,
        randomize: true
      });

      await Domain.createQueryBuilder('domain')
        .update(Domain)
        .set({ syncedAt: new Date(Date.now()) })
        .where({ id: In(domains.map((e) => e.id)) })
        .execute();
    }
    console.log('Domain sync complete.');
  } else {
    console.log('Not syncing any domains.');
  }
};
