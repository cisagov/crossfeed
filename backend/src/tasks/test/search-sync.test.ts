import { handler as searchSync, DOMAIN_CHUNK_SIZE } from '../search-sync';
import {
  connectToDatabase,
  Organization,
  Domain,
  Service,
  Vulnerability
} from '../../models';

jest.mock('../es-client');
const { updateDomains, updateWebpages } = require('../es-client');

describe('search_sync', () => {
  let organization;
  beforeAll(async () => {
    await connectToDatabase();
  });
  beforeEach(async () => {
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
  });

  test('should not update already-synced domains', async () => {
    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-10-10')
    }).save();

    await Service.create({
      service: 'https',
      port: 443,
      domain
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).not.toBeCalled();
  });

  test('should update a domain if a service has changed', async () => {
    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-10-10')
    }).save();

    await Service.create({
      service: 'https',
      port: 443,
      domain,
      updatedAt: new Date('9999-10-11')
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).toBeCalled();
  });

  test('should not update a domain if a service has not changed', async () => {
    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-10-10')
    }).save();

    await Service.create({
      service: 'https',
      port: 443,
      domain,
      updatedAt: new Date('9999-9-11')
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).not.toBeCalled();
  });

  test('should update a domain if a vulnerability has changed', async () => {
    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-10-10')
    }).save();

    await Vulnerability.create({
      domain,
      title: 'vuln',
      updatedAt: new Date('9999-10-11')
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).toBeCalled();
  });

  test('should not update a domain if a vulnerability has not changed', async () => {
    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-10-10')
    }).save();

    await Vulnerability.create({
      domain,
      title: 'vuln',
      updatedAt: new Date('9999-9-11')
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).not.toBeCalled();
  });

  test('should update a domain if an organization has changed', async () => {
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false,
      updatedAt: new Date('9999-10-11')
    }).save();

    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-10-10')
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).toBeCalled();
  });

  test('should not update a domain if an organization has not changed', async () => {
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false,
      updatedAt: new Date('9999-9-11')
    }).save();

    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-10-10')
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).not.toBeCalled();
  });

  test('should update a domain if a domain has changed', async () => {
    const domain = await Domain.create({
      name: 'cisa.gov',
      organization,
      syncedAt: new Date('9999-09-19T19:57:32.346Z'),
      updatedAt: new Date('9999-09-20T19:57:32.346Z')
    }).save();

    const service = await Service.create({
      service: 'https',
      port: 443,
      domain,
      updatedAt: new Date('9999-09-12T19:57:32.346Z')
    }).save();

    const vulnerability = await Vulnerability.create({
      title: 'test-' + Math.random(),
      domain
    }).save();

    await searchSync({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).toBeCalled();
    expect(
      Object.keys((updateDomains as jest.Mock).mock.calls[0][0][0])
    ).toMatchSnapshot();

    const domains = updateDomains.mock.calls[0][0];
    expect(domains.length).toEqual(1);
    expect(domains[0].id).toEqual(domain.id);
    expect(domains[0].organization.id).toEqual(organization.id);
    expect(domains[0].services.map((e) => e.id)).toEqual([service.id]);
    expect(domains[0].vulnerabilities.map((e) => e.id)).toEqual([
      vulnerability.id
    ]);

    const newDomain = await Domain.findOneOrFail(domain.id);
    expect(newDomain.syncedAt).not.toEqual(domain.syncedAt);
  });

  test('should sync domains in chunks', async () => {
    await Promise.all(
      new Array(DOMAIN_CHUNK_SIZE + 1).fill(null).map((e) =>
        Domain.create({
          name: 'cisa-' + Math.random() + '.gov',
          organization,
          syncedAt: new Date('9999-09-19T19:57:32.346Z'),
          updatedAt: new Date('9999-09-20T19:57:32.346Z')
        }).save()
      )
    );

    await searchSync({
      organizationId: organization.id,
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });

    expect(updateDomains).toBeCalledTimes(2);
  });
});
