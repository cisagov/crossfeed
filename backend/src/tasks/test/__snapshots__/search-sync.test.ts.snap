// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search_sync should update a domain if a domain has changed 1`] = `
Array [
  "id",
  "createdAt",
  "updatedAt",
  "syncedAt",
  "ip",
  "fromRootDomain",
  "reverseName",
  "name",
  "screenshot",
  "country",
  "asn",
  "cloudHosted",
  "ssl",
  "services",
  "organization",
  "vulnerabilities",
]
`;
